client/web/src/enterprise/batches/detail/changesets/ChangesetSelectRow.tsx

```diff
@@ -1,47 +1,240 @@
+import classNames from 'classnames'
 import InfoCircleOutlineIcon from 'mdi-react/InfoCircleOutlineIcon'
-import React from 'react'
+import React, { Fragment, useCallback, useMemo, useState } from 'react'
 
-import { isErrorLike } from '@sourcegraph/shared/src/util/errors'
 import { pluralize } from '@sourcegraph/shared/src/util/strings'
 
-import { ErrorAlert } from '../../../../components/alerts'
+import { AllChangesetIDsVariables, Scalars } from '../../../../graphql-operations'
+import { eventLogger } from '../../../../tracking/eventLogger'
+import { queryAllChangesetIDs } from '../backend'
+
+import styles from './ChangesetSelectRow.module.scss'
+import { CreateCommentModal } from './CreateCommentModal'
+import { DetachChangesetsModal } from './DetachChangesetsModal'
+
+/**
+ * Describes a possible action on the changeset list.
+ */
+interface ChangesetListAction {
+    /* The type of action. Used internally. */
+    type: string
+    /* The button label for the action. */
+    buttonLabel: string
+    /* The title in the dropdown menu item. */
+    dropdownTitle: string
+    /* The description in the dropdown menu item. */
+    dropdownDescription: string
+    /* Conditionally display the action based on the given query arguments. */
+    isAvailable: (queryArguments: Omit<AllChangesetIDsVariables, 'after'>) => boolean
+    /**
+     * Invoked when the action is triggered. Either onDone or onCancel need to be called
+     * eventually. Can return a JSX.Element to be rendered adacent to the button (i.e. a modal).
+     */
+    onTrigger: (
+        batchChangeID: Scalars['ID'],
+        changesetIDs: () => Promise<Scalars['ID'][]>,
+        onDone: () => void,
+        onCancel: () => void
+    ) => void | JSX.Element
+}
+
+const AVAILABLE_ACTIONS: ChangesetListAction[] = [
+    {
+        type: 'detach',
+        buttonLabel: 'Detach changesets',
+        dropdownTitle: 'Detach changesets',
+        dropdownDescription:
+            "Removes the selected changesets from this batch change. Unlike archive, this can't be undone.",
+        // Only show on the archived tab.
+        isAvailable: ({ onlyArchived }) => !!onlyArchived,
+        onTrigger: (batchChangeID, changesetIDs, onDone, onCancel) => (
+            <DetachChangesetsModal
+                batchChangeID={batchChangeID}
+                changesetIDs={changesetIDs}
+                afterCreate={onDone}
+                onCancel={onCancel}
+                telemetryService={eventLogger}
+            />
+        ),
+    },
+    {
+        type: 'commentatore',
+        buttonLabel: 'Create comment',
+        dropdownTitle: 'Create comment',
+        dropdownDescription:
+            'Create a comment on all selected changesets to ask people for reviews, or give an update.',
```

Comment:
This feels a little closed off and formal:

```suggestion
            'Create a comment on all selected changesets. For example, you could ask people for reviews, give an update, or post a cat GIF.',
```